hazelcast/src/test/java/com/hazelcast/client/cache/impl/nearcache/ClientCacheNearCacheCacheOnUpdateTest.java

```diff
@@ -0,0 +1,141 @@
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import javax.cache.spi.CachingProvider;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE_ON_UPDATE;
+import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.INVALIDATE;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ClientCacheNearCacheCacheOnUpdateTest extends ClientNearCacheTestSupport {
+
+    private static final int NUM_OF_KEYS = 100;
+
+    @Test
+    public void with_cacheOnUpdate_policy_concurrently_updated_near_cache_does_not_cause_any_miss()
+            throws InterruptedException {
+        ICache<Integer, Integer> nearCachedClientCache = newNearCachedCache(CACHE_ON_UPDATE);
+
+        runTest(nearCachedClientCache, CACHE_ON_UPDATE);
+
+        assertEquals(0, nearCachedClientCache.getLocalCacheStatistics()
+                .getNearCacheStatistics().getMisses());
+    }
+
+    @Test
+    public void with_invalidate_policy_concurrently_updated_near_cache_causes_misses()
+            throws InterruptedException {
+        ICache<Integer, Integer> nearCachedClientCache = newNearCachedCache(INVALIDATE);
+
+        runTest(nearCachedClientCache, INVALIDATE);
+
+        assertTrue(nearCachedClientCache.getLocalCacheStatistics()
+                .getNearCacheStatistics().getMisses() > NUM_OF_KEYS);
+    }
+
+    private static void runTest(ICache<Integer, Integer> icacheOnClient,
+                                NearCacheConfig.LocalUpdatePolicy localUpdatePolicy) throws InterruptedException {
+        for (int i = 0; i < NUM_OF_KEYS; i++) {
+            icacheOnClient.put(i, i);
+
+            if (localUpdatePolicy.equals(INVALIDATE)) {
+                // populate near-cache here for invalidate policy.
+                icacheOnClient.get(i);
+            }
+        }
+
+        AtomicBoolean stop = new AtomicBoolean(false);
+
+        Runnable getter = () -> {
+            while (!stop.get()) {
+                for (int i = 0; i < NUM_OF_KEYS; i++) {
+                    icacheOnClient.get(i);
+                }
+            }
+        };
+
+        Runnable putter = () -> {
+            while (!stop.get()) {
+                for (int i = 0; i < NUM_OF_KEYS; i++) {
+                    icacheOnClient.put(i, i);
+                }
+            }
+        };
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+
+        int numOfGetters = 3 * RuntimeAvailableProcessors.get();
+        for (int i = 0; i < numOfGetters; i++) {
+            executor.execute(getter);
+        }
+
+        executor.execute(putter);
+
+        sleepSeconds(20);
```

Comment:
Not sure that we can consider the tests as `QuickTest`

Reply:
I want it to run as much as possible within PR builders. For now decreased to 10 which is still a long time. Later it can be decreased or mark as Slow.